### ISSUE TITLE
create klp_patch struct at build time

### ISSUE THREAD

**Issue by jpoimboe** [dynup/kpatch#478]
Create the klp_patch struct at build time, rather than run time, as discussed in #477.  This will greatly simplify the patch module template code.

**Comment by arges** [comment]
I've started hacking on this. How does this sound:

1) Modify create-diff-object.c with an '-l' flag for livepatch. This flag will output .klp.funcs, .klp.relocs that match klp_* structures (much like kpatch-patch.h).

2) Modify kpatch-build to invoke create-diff-object with -l when klp patching is detected.

3) Create a program that links all patch objects and inserts constructed klp_objects. Each klp_object references its own .klp.<objectname>.relocs and .klp.<objectname>.funcs sections.

4) The livepatch hook script constructs the klp_patch and simply inserts the linked klp_objects.

**Comment by jpoimboe** [comment]
@arges Sounds great.  Though I didn't quite follow this:

> 3) Create a program that links all patch objects and inserts constructed klp_objects. Each klp_object references its own .klp..relocs and .klp..funcs sections.

Why create a program?  Doesn't kpatch-build do the object linking already?  (I must admit I haven't looked at the code in quite a while, so I could be missing something obvious...)

**Comment by arges** [comment]
A few challenges I'm trying to figure out:

1)
The create-diff-object program outputs funcs/relocs for one klp_object. However creating a section like .klp.funcs / .klp.relocs would require an additional field (like objname) to correlate them an object later on (like how .kpatch.funcs already does this). However doing this would mean the livepatch hook will still need to dynamically create the structures.

I'm trying to figure out how to do this with just a linker and templating such that the livepatch hook code externs the variables from the linked object and registers the structures. One option would be to create the klp_object(s) in create-diff-object which already have pre-populated indicies for funcs/relocs into .klp.funcs/.klp.relocs. I'd have to ensure that there are zero'ed elements in between each.

2)
Including the livepatch.h header in create-diff-objects from linux/livepatch.h would be ideal. However, this is challenging because of the list_head conflicts etc. Currently I have a version of this header and put it into kmod/patch/klp_patch.h, such that the 'internal' fields are just char _pad[X]. My worry here is these structures getting out of sync with the kernel (or other CONFIG discrepancies.)

3) 
Separating out kpatch/klp code. Right now I can write lots of if statements and pass flags and what not. Is there a better way to handle this? I could also #ifdef code if it makes things easier.

I'll continue to hack, let me know any advise here.
Thanks,

**Comment by jpoimboe** [comment]
> 1)
> The create-diff-object program outputs funcs/relocs for one klp_object. However creating a section like .klp.funcs / .klp.relocs would require an additional field (like objname) to correlate them an object later on (like how .kpatch.funcs already does this). However doing this would mean the livepatch hook will still need to dynamically create the structures.
> 
> I'm trying to figure out how to do this with just a linker and templating such that the livepatch hook code externs the variables from the linked object and registers the structures. One option would be to create the klp_object(s) in create-diff-object which already have pre-populated indicies for funcs/relocs into .klp.funcs/.klp.relocs. I'd have to ensure that there are zero'ed elements in between each.

Yeah, this sounds pretty hard.  I guess you'd have to create the klp_object in create-diff-object, and also have the linker script be smart enough to combine multiple klp_objects into a single one if they have the same object name (e.g. "vmlinux").  My linker-fu is rusty, so I don't know if that's possible or not.

> 2)
> Including the livepatch.h header in create-diff-objects from linux/livepatch.h would be ideal. However, this is challenging because of the list_head conflicts etc. Currently I have a version of this header and put it into kmod/patch/klp_patch.h, such that the 'internal' fields are just char _pad[X]. My worry here is these structures getting out of sync with the kernel (or other CONFIG discrepancies.)

Oh, right.  This is tricky.  I guess ideally there would be a uapi version of livepatch.h with some more user space friendly interfaces.  But since the livepatch user space story is still vague, I guess the best we can do for now is just duplicate them and manually keep them in sync.

> 3)
> Separating out kpatch/klp code. Right now I can write lots of if statements and pass flags and what not. Is there a better way to handle this? I could also #ifdef code if it makes things easier.

One thing that I think would make this easier is changing the kpatch interfaces to more closely match the klp interfaces.  Ideally the kpatch core module would have register() functions and structs and sysfs entries which are as close to the klp_* interfaces as possible.  Hopefully that would reduce the number of places in the code where they diverge.

BTW, if this change ends up requiring too much hacking, or results in a lot of spaghetti code, we might want to reassess whether it's a good idea.  i.e. hopefully in the process of simplifying the run time code, we don't over-complicate the build time code (which is already quite complex).

**Comment by arges** [comment]
On 10/19/2015 01:04 PM, Josh Poimboeuf wrote:
>     1)
>     The create-diff-object program outputs funcs/relocs for one
>     klp_object. However creating a section like .klp.funcs / .klp.relocs
>     would require an additional field (like objname) to correlate them
>     an object later on (like how .kpatch.funcs already does this).
>     However doing this would mean the livepatch hook will still need to
>     dynamically create the structures.
> 
>     I'm trying to figure out how to do this with just a linker and
>     templating such that the livepatch hook code externs the variables
>     from the linked object and registers the structures. One option
>     would be to create the klp_object(s) in create-diff-object which
>     already have pre-populated indicies for funcs/relocs into
>     .klp.funcs/.klp.relocs. I'd have to ensure that there are zero'ed
>     elements in between each.
> 
> Yeah, this sounds pretty hard. I guess you'd have to create the
> klp_object in create-diff-object, and also have the linker script be
> smart enough to combine multiple klp_objects into a single one if they
> have the same object name (e.g. "vmlinux"). My linker-fu is rusty, so I
> don't know if that's possible or not.
> 

So thinking about this a bit more. Here are some approaches:

1) Write something similar to what the livepatch-patch-hook.c function
already does by associating all objname's to the correct objects and
dynamically constructing the right structures. Dump this out into
another object file along with the appropriate sections. Seems hackish.

2) Each object file passed into create-diff-object can write
.livepatch.funcs.<objname> (matching the final klp_func structures)
which get merged during linking. Then with some Makefile/Bash script
magic we generate a linker template that does the correct thing (oh and
probably have to tell the module code a list of the objects by name).
The module hook still has to construct the patch and objects.

3) Modify create-diff-object.c to iterate on all object files at once.
This would allow funcs/relocs sections to be merged per object and
output all at once without additional steps. However, I'm still wrapping
my head around this code, and not sure how feasible it is.
This seems to be the cleanest, but would fundamentally change this code
(maybe it will make kpatch-patch construction easier too).

Any thoughts here would be helpful. I did (2) with just a single vmlinux
object, but doing this with multiple objnames and multiple object files
would make this much more complex at the bash/makefile/linker script side.

--chris

**Comment by jpoimboe** [comment]
> 1) Write something similar to what the livepatch-patch-hook.c function already does by associating all objname's to the correct objects and dynamically constructing the right structures. Dump this out into another object file along with the appropriate sections. Seems hackish.

This is an interesting idea.  It does add another step to the build process, but it has some benefits.  It seems more modular than the other approaches.  It encapsulates knowledge about the layout of the klp_* struct away from create-diff-object, and also keeps create-diff-object.c leaner.

I think it would also mean that create-diff-object wouldn't need to know about kpatch vs livepatch, right?  That would be another win.  And if we changed the kpatch core module structs to be more klp-esque, maybe this new tool could easily handle both kpatch and livepatch.

To avoid duplicating all the elf read/write code needed by this new tool, we could move all the elf-specific stuff from create-diff-object.c into elf.c and elf.h, so that they could be used by both tools.  (In fact I already did something similar with stacktool.)

> 2) Each object file passed into create-diff-object can write .livepatch.funcs.<objname> (matching the final klp_func structures) which get merged during linking. Then with some Makefile/Bash script magic we generate a linker template that does the correct thing (oh and probably have to tell the module code a list of the objects by name). The module hook still has to construct the patch and objects.

I originally wondered if this could be done by the linker script only (without creating a template + bash magic), but it looks like that's not possible.  So now this is probably my least favorite option.

> 3) Modify create-diff-object.c to iterate on all object files at once. This would allow funcs/relocs sections to be merged per object and output all at once without additional steps. However, I'm still wrapping my head around this code, and not sure how feasible it is. This seems to be the cleanest, but would fundamentally change this code (maybe it will make kpatch-patch construction easier too).

This could work.  But create-diff-object already does so much that I'm a little uneasy about enlarging its scope.

@spartacus06, @flaming-toast, any thoughts?

**Comment by jpoimboe** [comment]
Another benefit of option 1 is that we might someday want to make create-diff-object more generic so that it works with _any_ object files, not just the Linux kernel.  That would be another good reason to leave the klp-specific stuff out of create-diff-object.

So I would vote option 1.

**Comment by arges** [comment]
Hi,
I've finally started implementing this and have a non-functional patchset. I'd appreciate some advice as I'm getting a little stuck. I'll explain my process to help give some context.

After creating the output.o object from `create-diff-object`, instead of linking this into the final kernel module this gets passed into `create-livepatch-object`.  The `create-livepatch-object` program then uses some of the common elf functions to open the file and manipulate and read the elf sections. The `.kpatch.funcs` and `.kpatch.dynrelas` sections are read and put into lists in order to sort through which objects they actually patch. (This logic was copied from the livepatch hook.) After these are sorted we then iterate through the list and create actual `klp_*` structures appropriate for livepatch. These
are then written out to `.livepatch.*` sections. The module hook then just links against this output object and reads in the structure as assigns the objects to the dynamically allocated `klp_patch` object.

This doesn't work, and I think where I'm being hung up on is not re-doing where the relocations are in the output elf. I'm not sure how/when I should be translating these relocations and when I should be re-writing these. I'm trying hard to think of ways to not re-invent things. Suggestions?

My working (non-functional, but it does compile and produce output) branch:
https://github.com/arges/kpatch/tree/klp_static

--chris j arges

**Comment by jpoimboe** [comment]
@arges, sorry for taking so long to respond.  I got pulled in other directions and completely missed your comment.  I think the relocations will be tricky to deal with.

I feel bad for saying this because I know you put a lot of effort into this.  But I'm thinking this isn't such a good idea after all.  I agree with my original self when I said that build complexity is better than runtime complexity, but as it turns out, I think the build code would end up being *much* more complex than the patch hook code, and I don't think the tradeoff is worth it.

And the patch hook code is pretty straightforward after all: just allocating and copying structures.  Also, another advantage of doing it in the patch hook code is that I think we can more easily deal with backward compatibility by testing the kernel version with ifdefs.

I'm really sorry if I wasted your time.  It really did seem like a good idea before I realized all the details involved...
